ssl/s3_lib.c

```diff
@@ -3770,6 +3769,34 @@ OPENSSL_GLOBAL const SSL_CIPHER ssl3_ciphers[] = {
      256,
      },
 
+    {
+     1,
+     "GOST2012-GOST8912-GOST8912",
+     0x0300ff85,
+     SSL_kGOST,
+     SSL_aGOST12 | SSL_aGOST01,
+     SSL_eGOST2814789CNT12,
+     SSL_GOST89MAC12,
+     SSL_TLSV1,
+     SSL_NOT_EXP | SSL_HIGH,
+     SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC,
```

Comment:
Here is the patch against my fork (https://github.com/beldmit/openssl/tree/tls_gost_2012)
--- a/ssl/s3_enc.c
+++ b/ssl/s3_enc.c
@@ -537,7 +537,8 @@ int ssl3_digest_cached_records(SSL *s, int keep)
         }
 
         /* Loop through bits of algorithm2 field and create MD_CTX-es */
-        for (i = 0; ssl_get_certverify_digest(i, &mask, &md); i++) {
+        /* for (i = 0; ssl_get_certverify_digest(i, &mask, &md); i++) { */ 
+        for (i = 0; ssl_get_handshake_digest(i, &mask, &md); i++) {
             if ((mask & ssl_get_algorithm2(s)) && md) {
                 s->s3->handshake_dgst[i] = EVP_MD_CTX_create();
                 if (EVP_MD_nid(md) == NID_md5) {
diff --git a/ssl/s3_srvr.c b/ssl/s3_srvr.c
index ebbdebd..08c0e2e 100644
--- a/ssl/s3_srvr.c
+++ b/ssl/s3_srvr.c
@@ -579,7 +579,7 @@ int ssl3_accept(SSL *s)
                  */
                 s->state = SSL3_ST_SR_CHANGE_A;
                 s->init_num = 0;
-            } else if (SSL_USE_SIGALGS(s)) {
+            } else if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
                 s->state = SSL3_ST_SR_CERT_VRFY_A;
                 s->init_num = 0;
                 if (!s->session->peer)
@@ -1427,7 +1427,7 @@ int ssl3_get_client_hello(SSL *s)
         s->s3->tmp.new_cipher = s->session->cipher;
     }
 
-    if (!SSL_USE_SIGALGS(s) || !(s->verify_mode & SSL_VERIFY_PEER)) {
+    if (!(SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) ) )|| !(s->verify_mode & SSL_VERIFY_PEER)) {
         if (!ssl3_digest_cached_records(s, 0))
             goto f_err;
     }
@@ -2832,7 +2832,7 @@ int ssl3_get_cert_verify(SSL *s)
     if (n == 64 && (pkey->type == NID_id_GostR3410_2001)) {
          len = 64;
     } else {
-        if (SSL_USE_SIGALGS(s)) {
+        if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
             int rv;
 
             if (!PACKET_get_bytes(&pkt, &sig, 2)) {
@@ -2869,7 +2869,7 @@ int ssl3_get_cert_verify(SSL *s)
         goto f_err;
     }
 
-    if (SSL_USE_SIGALGS(s)) {
+    if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
         long hdatalen = 0;
         void *hdata;
         hdatalen = BIO_get_mem_data(s->s3->handshake_buffer, &hdata);


It breaks self-compatibility 
(
s_client -connect localhost:4433 -CAfile sslCA/cacert.pem -verify_return_error -verify 1 -state -cert ssl_user_gost2012_256_B/cert.pem -key ssl_user_gost2012_256_B/seckey.pem -ign_eof
s_server -msg -cert localhost_gost2001_XA/cert.pem -key localhost_gost2001_XA/seckey.pem -Verify 3 -verify_return_error -www -CAfile sslCA/cacert.pem -tls1

Diagnostics:
47070142607912:error:1414D172:SSL routines:tls12_check_peer_sigalg:wrong signature type:t1_lib.c:1017:
47070142607912:error:140940E5:SSL routines:ssl3_read_bytes:ssl handshake failure:record/rec_layer_s3.c:1034:
)

and compatibility with the 3rd-party implementation
(
s_client -msg -cipher GOST2012-GOST8912-GOST8912 -cert U_x_gost2001_B/cert.pem -key U_x_gost2001_B/seckey.pem -CAfile ca_gost2012_512.pem -verify_return_error -verify 1 -connect v-cp4-12L.vm.cryptocom.ru:443 -ign_eof

47777785954856:error:1411E144:SSL routines:tls1_cert_verify_mac:no required digest:t1_enc.c:709:
47777785954856:error:140940E5:SSL routines:ssl3_read_bytes:ssl handshake failure:record/rec_layer_s3.c:1260:
)